### PR TITLE
clientv3: fix deadlock on Get with concurrent Close

### PR DESCRIPTION
**Pull request by heyitsanthony**
/cc @gyuho

### CODE REVIEW

**Comment by gyuho** [comment]
ah, right. Thanks! LGTM.

File: clientv3/integration/kv_test.go [hunk 1]
```diff
@@ -288,14 +288,8 @@ func TestKVGetErrConnClosed(t *testing.T) {
...
 
-	closed, donec := make(chan struct{}), make(chan struct{})
+	donec := make(chan struct{})
 	go func() {
-		select {
-		case <-time.After(3 * time.Second):
-			t.Fatal("cli.Close took too long")
-		case <-closed:
-		}
-
```
Comment (gyuho) [inline clientv3/integration/kv_test.go]: Or how about just move `kv := clientv3.NewKV(cli)` below here?

```go
package main

import (
	"fmt"
	"log"
	"time"

	"golang.org/x/net/context"

	"github.com/coreos/etcd/clientv3"
)

var (
	dialTimeout    = 5 * time.Second
	requestTimeout = 1 * time.Second
	endpoints      = []string{"localhost:2379"}
)

func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: dialTimeout,
	})
	if err != nil {
		log.Fatal(err)
	}

	closed, donec := make(chan struct{}), make(chan struct{})
	// donec := make(chan struct{})
	go func() {
		select {
		case <-time.After(3 * time.Second):
			log.Fatal("cli.Close took too long")
		case <-closed:
		}

		kv := clientv3.NewKV(cli)

		// ctx, cancel := context.WithTimeout(cli.Ctx(), 3*time.Second) // cli.Close cancels this context, and will panic
		ctx, cancel := context.WithTimeout(context.TODO(), 3*time.Second) // cli.Close does not cancel this context, and still panic

		fmt.Println("kv.Get 1", ctx.Err())
		_, err := kv.Get(ctx, "test")
		fmt.Println("kv.Get 2")

		fmt.Println("cancel")
		cancel()
		fmt.Println("canceled")

		if err != nil {
			log.Fatal(err)
		}
		close(donec)
	}()

	if err := cli.Close(); err != nil {
		log.Fatal(err)
	}
	close(closed)

	select {
	case <-time.After(3 * time.Second):
		log.Fatal("kv.Get took too long")
	case <-donec:
	}
	fmt.Println("DONE")
}
```

But below doesn't panic without the previous patch:

```go
package main

import (
	"fmt"
	"log"
	"time"

	"golang.org/x/net/context"

	"github.com/coreos/etcd/clientv3"
)

var (
	dialTimeout    = 5 * time.Second
	requestTimeout = 1 * time.Second
	endpoints      = []string{"localhost:2379"}
)

func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: dialTimeout,
	})
	if err != nil {
		log.Fatal(err)
	}

	kv := clientv3.NewKV(cli)

	// closed, donec := make(chan struct{}), make(chan struct{})
	donec := make(chan struct{})
	go func() {
		// select {
		// case <-time.After(3 * time.Second):
		// 	log.Fatal("cli.Close took too long")
		// case <-closed:
		// }

		// kv := clientv3.NewKV(cli)

		// ctx, cancel := context.WithTimeout(cli.Ctx(), 3*time.Second) // cli.Close cancels this context, and will panic
		ctx, cancel := context.WithTimeout(context.TODO(), 3*time.Second) // cli.Close does not cancel this context, and still panic

		fmt.Println("kv.Get 1", ctx.Err())
		_, err := kv.Get(ctx, "test")
		fmt.Println("kv.Get 2")

		fmt.Println("cancel")
		cancel()
		fmt.Println("canceled")

		if err != nil {
			log.Fatal(err)
		}
		close(donec)
	}()

	if err := cli.Close(); err != nil {
		log.Fatal(err)
	}
	// close(closed)

	select {
	case <-time.After(3 * time.Second):
		log.Fatal("kv.Get took too long")
	case <-donec:
	}
	fmt.Println("DONE")
}

```


/cc @heyitsanthony

**Comment by gyuho** [comment]
And I tried this patch with the same code here and still panics?

```go
package main

import (
	"fmt"
	"log"
	"time"

	"golang.org/x/net/context"

	"github.com/coreos/etcd/clientv3"
)

var (
	dialTimeout    = 5 * time.Second
	requestTimeout = 1 * time.Second
	endpoints      = []string{"localhost:2379"}
)

func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: dialTimeout,
	})
	if err != nil {
		log.Fatal(err)
	}

	// kv := clientv3.NewKV(cli)

	closed, donec := make(chan struct{}), make(chan struct{})
	// donec := make(chan struct{})
	go func() {
		select {
		case <-time.After(3 * time.Second):
			log.Fatal("cli.Close took too long")
		case <-closed:
		}

		kv := clientv3.NewKV(cli)

		// ctx, cancel := context.WithTimeout(cli.Ctx(), 3*time.Second) // cli.Close cancels this context, and will panic
		ctx, cancel := context.WithTimeout(context.TODO(), 3*time.Second) // cli.Close does not cancel this context, and still panic

		fmt.Println("kv.Get 1", ctx.Err())
		_, err := kv.Get(ctx, "test")
		fmt.Println("kv.Get 2")

		fmt.Println("cancel")
		cancel()
		fmt.Println("canceled")

		if err != nil {
			log.Fatal(err)
		}
		close(donec)
	}()

	if err := cli.Close(); err != nil {
		log.Fatal(err)
	}
	close(closed)

	select {
	case <-time.After(3 * time.Second):
		log.Fatal("kv.Get took too long")
	case <-donec:
	}
	fmt.Println("DONE")
}
```

Or I am doing something wrong.

Can you try?

Thanks!

**Comment by heyitsanthony** [comment]
@gyuho can repro, investigating

**Comment by heyitsanthony** [comment]
@gyuho fixed the panic

**Comment by gyuho** [comment]
lgtm. thanks!

**Comment by xiang90** [comment]
@heyitsanthony Can you explain what is the cause of the deadlock?

**Comment by heyitsanthony** [comment]
@xiang90 `r.acquire()` returns holding `r.client.mu.RLock()` on success; it was dead locking because it was returning with the rlock held on a failure path and leaking it. After that any call to `client.Close()` will block forever waiting for the wlock.

**Comment by xiang90** [comment]
lgtm. Probably adding some comments to acquire and release would be helpful.

**Pull request merged**